Better arrows

JorenC opened the pull request:
Adding the PR now, rather than make it a huge thing. 
- Added a lot of setup to the stories so we can see more orders - this is a real game.
- Changed arrows so they use the colour of the origin nation.
- Changed initial values of the arrows
- Added different fail states
- Changes the '+' to 'x' on the orders (you passed "angle="angle" but I think it was nowhere defined, I now hardcoded it to 45)
- Included support hold order
- Included convoy order

johnpooch committed:
Merge pull request #12 from johnpooch/johnpooch/title-favicon

Update favicon and change document title to Diplicity

johnpooch committed:
Merge pull request #13 from johnpooch/johnpooch/phases

Add phase select functionality to game details screens

johnpooch committed:
Merge pull request #2 from johnpooch/johnpooch/add-readme

Update README to reflect project details and setup instructions

johnpooch committed:
Merge pull request #14 from johnpooch/johnpooch/phases

Fix issue with selectedPhase deleting all qs params

Commit:
Add order outcomes

Commit:
case

Commit:
Fix casing

johnpooch committed:
Merge pull request #16 from johnpooch/johnpooch/order-resolutions

Add order outcomes

Commit:
Move contexts to context sub-folder

Commit:
Refactor IconButton usage in PhaseSelect component to use StyledIconButton for consistent styling

Commit:
Add "Invalid" outcome to orders utility and update label and icon mappings

Commit:
wip

Commit:
Add utility functions for order formatting and resolution transformation; implement tests and update order handling

Commit:
wip

Commit:
Refactor game-related hooks and components; remove unused files and consolidate logic in useFindGames and useMyGames

Commit:
wip

Commit:
Add styles for tabs in My Games screen to ensure full-width display

Commit:
Refactor Login screen layout; replace Card with Stack for improved structure and add logo image; update My Games image size

Commit:
Implement player info and game info

Commit:
Add Google Fonts and update theme typography to use 'Cabin' font

Commit:
wip

Commit:
wip

Commit:
Refactor game detail screen: separate Map and Orders components, remove MapOrders

Commit:
Enhance orders display: add AppBar with navigation and group orders by nation

Commit:
Add GameDetailAppBar component and integrate it into Map and Orders screens

Commit:
Update navigation items to use dynamic gameId for routing in GameDetailLayout

Commit:
Add game-info and player-info routes; enhance GameDetailAppBar with navigation menu

Commit:
Enhance PhaseSelect component for mobile responsiveness; update GameDetailAppBar with a divider and improve orders display message

Commit:
Add Chat screen and integrate it into routing; create MapOrders component for displaying orders alongside the game map

Commit:
Add OrderList component and use it in Orders screen; refactor useOrders hook for improved order management

Commit:
update favicon

Commit:
Update favicon

Commit:
Refactor game-related schemas and enhance GameCard component with join/leave functionality

Commit:
Refactor CreateOrder hook and enhance Orders screen with a bottom sheet for order creation

Commit:
Add message and channel schemas; update service to handle messages and channels

Commit:
Refactor CreateOrder component; add ChannelList component and update routing

Commit:
Add chat

Commit:
Add desktop chat

Commit:
Add styled desktop chat

Commit:
Refactor GameCard component to streamline navigation logic and update map styling

Commit:
Add telemetry service and update variant handling in order utilities

Commit:
Add telemetry service for error logging and update middleware to track rejected actions

Commit:
Refactor login screen structure and add CreateOrder component for order management

Commit:
Remove chat components and add GameDetailMenu for improved game detail navigation

Commit:
Add useGetUserPhaseStateQuery hook and integrate it into order management for enhanced phase state handling

Commit:
Refactor game schema to update ActiveBans structure for improved user and owner tracking

Commit:
Add IMapEditor interface for map editing functionality

Commit:
restructure layout

Commit:
s

Commit:
f

Commit:
s

Commit:
t

Commit:
t

Commit:
Add web.config for URL rewriting to support React routes

Commit:
Replace web.config with staticwebapp.config.json for React route handling

Commit:
Add spacing component to HomeLayout for mobile view

Commit:
Refactor GameCard component layout for improved readability and structure

Commit:
Increase token duration in login URL to 100 years for extended session management

Commit:
Add CreateChannelAction component and update channel-related exports

Commit:
Add interactive map

Commit:
Wip

johnpooch committed:
Merge pull request #28 from johnpooch/johnpooch/interactive-map

Johnpooch/interactive map

JorenC committed:
Improved Cross

Made the cross diagonal, better size/aspect ratio.

JorenC committed:
Adding units

JorenC committed:
Arrows now work

- Added many stories to a realistic map (this is real game)
- Arrows are now the colour of their origin unit

JorenC committed:
Added support order

JorenC committed:
Update SC rendering

Much better to see

JorenC committed:
Better Octagon colour

colour updated for the support octagon

JorenC committed:
Added Convoy Arrow

johnpooch commented on src/components/interactive-map/interactive-map.tsx:
```diff
@@ -64,7 +69,7 @@ const InteractiveMap: React.FC<InteractiveMapProps> = (props) => {
         /rgb(a?)\((\d+), (\d+), (\d+)(, [\d.]+)?\)/,
         // "rgba($2, $3, $4, 0.4)"
-        `rgba($2, $3, $4, ${isSelected ? 0.3 : isHovered ? 0.4 : 0.5})`
+        `rgba($2, $3, $4, ${isSelected ? 0.3 : isHovered ? 0.4 : 0.5})`,
```
Tiny thing, and we can pick this up in a future iteration, but this `0.3`, `0.4`, and `0.5` should be extracted into named constants like `PROVINCE_COLOR_OPACITY_HOVERED`.

johnpooch commented on src/components/interactive-map/interactive-map.tsx:
```diff
@@ -143,18 +148,18 @@ const InteractiveMap: React.FC<InteractiveMapProps> = (props) => {
+                  fill="white"
                   stroke="black"
-                  strokeWidth={1}
+                  strokeWidth={2}
```
Can we extract all of these values into named constants:
- `SUPPLY_CENTER_SIZE_INNER`
- `SUPPLY_CENTER_SIZE_OUTER` 
- `SUPPLY_CENTER_FILL`
- `SUPPLY_CENTER_STROKE`
- `SUPPLY_CENTER_STROKE_WIDTH`

johnpooch commented on src/components/interactive-map/interactive-map.tsx:
```diff
@@ -263,8 +268,35 @@ const InteractiveMap: React.FC<InteractiveMapProps> = (props) => {
           const source = props.map.provinces.find((p) => p.id === provinceId);
           const target = props.map.provinces.find((p) => p.id === order.target);
+          const unit = props.units[provinceId];
+          const unitColor = unit ? props.nationColors[unit.nation] : "black"; // Default to black if no unit found
```
Curious, can this happen in practice?

JorenC replied:
I don't think really, unless it's a bug - but it helps identify issues when using StoryBook and is a nice fallback

johnpooch commented on src/components/interactive-map/interactive-map.tsx:
```diff
@@ -292,7 +324,7 @@ const InteractiveMap: React.FC<InteractiveMapProps> = (props) => {
                         width={ORDER_FAILED_CROSS_WIDTH}
                         length={ORDER_FAILED_CROSS_LENGTH}
-                        angle={angle}
+                        angle={45}
```
Yeah, I saw your comment about this. So the `angle` value that gets passed to this callback is the angle of the arrow being drawn on, so my thinking was we would either do `angle`  or `angle + 45`  or something. But if what you have works, that's fine.

JorenC replied:
Let's leave it at this then.

johnpooch commented on src/components/interactive-map/interactive-map.tsx:
```diff
@@ -303,12 +335,15 @@ const InteractiveMap: React.FC<InteractiveMapProps> = (props) => {
           const source = props.map.provinces.find((p) => p.id === provinceId);
           const target = props.map.provinces.find((p) => p.id === order.target);
+          const unit = props.units[provinceId];
+          const unitColor = unit ? props.nationColors[unit.nation] : "black"; // Default to black if no unit found
```
I'll do a subsequent iteration to tidy up some of this repeated code and stuff. I think I can restructure this file a little more nicely.

johnpooch commented on src/components/interactive-map/interactive-map.tsx:
```diff
@@ -320,9 +355,58 @@ const InteractiveMap: React.FC<InteractiveMapProps> = (props) => {
+              stroke={order.outcome === "failed" ? FAILED_COLOR : SUCCESS_COLOR}
+              fill={unitColor}
+              dash={{
+                length: ORDER_DASH_LENGTH * 2,
```
I would extract this into a named constant.

johnpooch commented on src/components/interactive-map/shapes/convoy-arrow.tsx:
```diff
@@ -0,0 +1,116 @@
+        cy={endY}
+        r={5} // Radius of the circle, can adjust this value
+        fill={"white"}
+        stroke={"black"}
```
I would pass these three values in as props, or maybe better, do an `onRenderEnd` callback where the parent component can pass the circle shape (similar to what we do with cross).

I think it's a nice pattern that keeps all the configurable stuff in the one file, making these shape components non-configurable.

johnpooch commented on src/components/interactive-map/shapes/convoy-arrow.tsx:
```diff
@@ -0,0 +1,116 @@
+  const angle = Math.atan2(endY - startY, endX - startX);
+
+  const centerX = (startX + endX) / 2;
+  const centerY = (startY + endY) / 2;
```
This is really nice stuff. Can't believe you wrote this _all_ on your own with _no_ help!

johnpooch commented on src/components/interactive-map/shapes/curved-arrow.tsx:
```diff
@@ -109,7 +109,7 @@ const CurvedArrow: React.FC<CurvedArrowProps> = (props) => {
   };
 
-  const centerPoint = getBezierPoint(0.5, start, cp1, cp2, end);
+  const centerPoint = getBezierPoint(0.3, start, cp1, cp2, end);
```
Is this something like the "amount" of curve? I think we should pass it in as a prop.

JorenC replied:
No, this is where the 'x' gets drawn on the support curve. 

If we do this at 0.5, this means that it usually overlaps with the actual move order. If we set it at about 0.3, it usually is where the support order does not yet overlap the actual move order. We could pass it as a prop, but I don't think we'll ever change this.

johnpooch commented on src/components/interactive-map/shapes/hold-arrow.tsx:
```diff
@@ -0,0 +1,94 @@
+  );
+};
+
+export { HoldArrow };
```
Let's rename this component to be just `Line`/`line.tsx`. Let's give the responsibility of rendering the octagon for the supported unit to `interactive-map.tsx`

This `shapes` subfolder is really about primitive shapes, and shouldn't have game logic built in.

Also, it saves us having to pass in the octagon values as props, which will make this component more messy.

So basically, for every "support hold" order, draw a `Line` component and an `Octagon`  component

johnpooch commented on src/components/interactive-map/shapes/convoy-arrow.tsx:
```diff
@@ -0,0 +1,116 @@
+  onRenderCenter?: (x: number, y: number, angle: number) => React.ReactElement;
+};
+
+const ConvoyArrow = (props: ConvoyArrowProps) => {
```
I have a question mark in my head about this component.

I think the `shapes` should all be primitive shapes and not have game logic built in.

I actually think this component could just be a `Line` component. We can get the point to which the line goes in a separate helper function:
``` 
const { endX, endY } = getClosestPointOnLine(x1, y1, x2, y2, x3, y3)
return (
  <Line x1={x1} y1={y1} x2={endX} y2={endY} />
)
```

JorenC replied:
This makes sense, but my plan is to use different lines. Right now the convoy line is straight, but i still want it to be wavey in the future. I'm okay with making primitives, although it's just one schema. The risk is combining different primitives will make it a bit harder to understand what is going on. I might take a look at this tomorrow if I have time

johnpooch reviewed (commented):
@JorenC I'll take a look at this tomorrow or the day after and implement the feedback I've left as well as doing some other general tidy-up.

JorenC committed:
Tokenize values

Tokenized the values. I'll try to keep this best practice - still not in my system as I design using the interface and then just believe that should be the value :D
